### PR TITLE
Added support for multi-object training and spatial embeddings

### LINKED ISSUE #31: Modify Base code to enable training multiple objects and spatial embeddings on InfoNCE loss

### PR DESCRIPTION
**Pull request by sakshamjindal**
Changes 
- Scene Graph - has a mode enabled. If mode==node, it returns a list of features [tensor(N_0,256), tensor(N_1,256), ... ] and encodings  are in the form(tensor : Nx256, where N is the number of objs in that batch element) 
- Encoder : Encoder returns a list of feature tensors and is a wrapper around the scene graph module
- MoCo model : input is the above list : Following are the operations performed
`output_k = encoder_k(feed_k)`  
`output_q = encoder_q(feed_q)`

We assume that the bounding boxes from the detector are unordered; we need to find the nearest neighour bbox pairs (object matching pair in the key and query. The following operation happens iteratively for each batch element (image) pair in key and query
`output_k, output_q = pair_embeddings(output_k, output_q)`

Now, node features in the entire batch with each batch element assumed having different number of boxes, are stacked to form a tensor [num_objects_in_entire_batch, 256]
`output_k = stack_embeddings(output_k)`
`output_q = stack_embeddings(output_q)`

Similar generalisation for spatial embeddings as well

### CODE REVIEW

File: core/model/model.py [hunk 1]
```diff
@@ -168,18 +169,25 @@ def forward(self, feed_dict_q, feed_dict_k=None, metadata=None, is_eval=False, c
...
 #             # undo shuffle
 #             k = self._batch_unshuffle_ddp(k, idx_unshuffle)
 
         # compute query features
-        _, q = self.encoder_q(feed_dict_q, mode)  # queries: NxC
-        q = nn.functional.normalize(q, dim=1)
+        q = self.encoder_q(feed_dict_q, mode)  # queries: NxC
+        #q = nn.functional.normalize(q, dim=1)
+
+        k,q = pair_embeddings(k,q)
```
Comment (ishitamed19) [inline core/model/model.py:182]: what does this function do for node embeddings? I mean, why do we need paired embedding if mode=="node"?

File: core/model/model.py [hunk 2]
```diff
@@ -145,14 +142,18 @@ def forward(self, feed_dict_q, feed_dict_k=None, metadata=None, is_eval=False, c
...
 
-
         if is_eval:
-            _, k = self.encoder_k(feed_dict_q, mode)
+            # the output from encoder is a list of features from the batch where each batch element (image)
+            # might contain different number of objects
+            k = self.encoder_k(feed_dict_q, mode)
+
+            # encoder output features in the list are stacked to form a tensor of features across the batch
+            k = stack_features_across_batch(k)
```
Comment (ishitamed19) [inline core/model/model.py:154]: what does this return - node embeddings or spatial embedding or both? or is it mode dependent?

File: core/model/utils.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+from sklearn.neighbors import NearestNeighbors
+import torch
+import torch.nn as nn
+
+def pair_embeddings(output_k, output_q, mode = "node"):
```
Comment (ishitamed19) [inline core/model/utils.py:5]: why is this funnction needed for node embedddings?

File: core/model/model.py (hunk 1 above)
Comment (sakshamjindal) [reply core/model/model.py:182]: Mode defines for the Scene graph module. If mode=='node', the scene graph module will return only node embeddings to the encoder Check scene_graph.py once.

File: core/model/model.py (hunk 2 above)
Comment (sakshamjindal) [reply core/model/model.py:154]: Depends on mode you are in. If it is 'node' it will return node embeddings only. If 'spatial', it will return only 'spatial embeddings' only

File: core/model/utils.py (hunk 3 above)
Comment (sakshamjindal) [reply core/model/utils.py:5]: Yeah. We can remove the mode from here. Not needed. Forgot to remove the mode here.
